<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Exception Handling</title>
    <meta name="generator" content="VuePress 1.8.2">
    
    <meta name="description" content="">
    
    <link rel="preload" href="/assets/css/0.styles.e4d8b86b.css" as="style"><link rel="preload" href="/assets/js/app.885421c5.js" as="script"><link rel="preload" href="/assets/js/2.63033d72.js" as="script"><link rel="preload" href="/assets/js/102.ad03aa52.js" as="script"><link rel="prefetch" href="/assets/js/10.0b3c205e.js"><link rel="prefetch" href="/assets/js/100.96b937c1.js"><link rel="prefetch" href="/assets/js/101.307416f3.js"><link rel="prefetch" href="/assets/js/103.f62a5b6a.js"><link rel="prefetch" href="/assets/js/104.af54bf4a.js"><link rel="prefetch" href="/assets/js/105.7e28a867.js"><link rel="prefetch" href="/assets/js/106.f43671e2.js"><link rel="prefetch" href="/assets/js/107.193e626f.js"><link rel="prefetch" href="/assets/js/108.48f6a137.js"><link rel="prefetch" href="/assets/js/109.cd1a3844.js"><link rel="prefetch" href="/assets/js/11.44430f62.js"><link rel="prefetch" href="/assets/js/110.1dd502da.js"><link rel="prefetch" href="/assets/js/111.ccb021d8.js"><link rel="prefetch" href="/assets/js/112.75e535d0.js"><link rel="prefetch" href="/assets/js/113.bfe1eaad.js"><link rel="prefetch" href="/assets/js/114.521ecc45.js"><link rel="prefetch" href="/assets/js/115.848462f3.js"><link rel="prefetch" href="/assets/js/116.f7937681.js"><link rel="prefetch" href="/assets/js/117.d00e5088.js"><link rel="prefetch" href="/assets/js/118.fd0e9def.js"><link rel="prefetch" href="/assets/js/119.c64902aa.js"><link rel="prefetch" href="/assets/js/12.d9d3b76c.js"><link rel="prefetch" href="/assets/js/120.c029eaac.js"><link rel="prefetch" href="/assets/js/121.17f7bf8c.js"><link rel="prefetch" href="/assets/js/122.a46a015c.js"><link rel="prefetch" href="/assets/js/123.bd8273f5.js"><link rel="prefetch" href="/assets/js/124.f0266688.js"><link rel="prefetch" href="/assets/js/125.23db0445.js"><link rel="prefetch" href="/assets/js/126.6c8e4b81.js"><link rel="prefetch" href="/assets/js/127.58121e7c.js"><link rel="prefetch" href="/assets/js/128.e81625d0.js"><link rel="prefetch" href="/assets/js/129.2086e255.js"><link rel="prefetch" href="/assets/js/13.02f01f86.js"><link rel="prefetch" href="/assets/js/130.358aa726.js"><link rel="prefetch" href="/assets/js/131.2db58096.js"><link rel="prefetch" href="/assets/js/132.bcbdada5.js"><link rel="prefetch" href="/assets/js/133.2c908f82.js"><link rel="prefetch" href="/assets/js/134.eade1bde.js"><link rel="prefetch" href="/assets/js/135.b89936b5.js"><link rel="prefetch" href="/assets/js/136.e88a267d.js"><link rel="prefetch" href="/assets/js/137.61035549.js"><link rel="prefetch" href="/assets/js/138.f89fbe70.js"><link rel="prefetch" href="/assets/js/139.771cd71c.js"><link rel="prefetch" href="/assets/js/14.7f3bb0b0.js"><link rel="prefetch" href="/assets/js/140.34a95e86.js"><link rel="prefetch" href="/assets/js/141.ec95bcce.js"><link rel="prefetch" href="/assets/js/142.13e1ee8e.js"><link rel="prefetch" href="/assets/js/143.1aa42e4b.js"><link rel="prefetch" href="/assets/js/144.37aab2b7.js"><link rel="prefetch" href="/assets/js/145.be3bc861.js"><link rel="prefetch" href="/assets/js/146.e5d674e1.js"><link rel="prefetch" href="/assets/js/147.58aec8b2.js"><link rel="prefetch" href="/assets/js/148.2b890544.js"><link rel="prefetch" href="/assets/js/149.abf23b17.js"><link rel="prefetch" href="/assets/js/15.9ee435da.js"><link rel="prefetch" href="/assets/js/150.294e67e5.js"><link rel="prefetch" href="/assets/js/151.0a384fa7.js"><link rel="prefetch" href="/assets/js/152.fdcc22f9.js"><link rel="prefetch" href="/assets/js/153.9da97a5a.js"><link rel="prefetch" href="/assets/js/154.9d75aa68.js"><link rel="prefetch" href="/assets/js/155.cb639e66.js"><link rel="prefetch" href="/assets/js/156.bac15746.js"><link rel="prefetch" href="/assets/js/157.59912327.js"><link rel="prefetch" href="/assets/js/158.70b73dcc.js"><link rel="prefetch" href="/assets/js/159.e88d2e22.js"><link rel="prefetch" href="/assets/js/16.35e658e9.js"><link rel="prefetch" href="/assets/js/160.f139525d.js"><link rel="prefetch" href="/assets/js/161.a37a29b7.js"><link rel="prefetch" href="/assets/js/162.6f5d3003.js"><link rel="prefetch" href="/assets/js/163.14aa422f.js"><link rel="prefetch" href="/assets/js/164.4b33e948.js"><link rel="prefetch" href="/assets/js/165.fb54ffee.js"><link rel="prefetch" href="/assets/js/166.041ac020.js"><link rel="prefetch" href="/assets/js/167.afa2fc40.js"><link rel="prefetch" href="/assets/js/168.e8dd5ce6.js"><link rel="prefetch" href="/assets/js/169.c6f23778.js"><link rel="prefetch" href="/assets/js/17.75911907.js"><link rel="prefetch" href="/assets/js/170.845b5d59.js"><link rel="prefetch" href="/assets/js/171.7e3dcda3.js"><link rel="prefetch" href="/assets/js/172.2d115602.js"><link rel="prefetch" href="/assets/js/173.872cef97.js"><link rel="prefetch" href="/assets/js/174.2ba2ca5e.js"><link rel="prefetch" href="/assets/js/175.dc1b572f.js"><link rel="prefetch" href="/assets/js/176.7692d4c3.js"><link rel="prefetch" href="/assets/js/177.22d1ef39.js"><link rel="prefetch" href="/assets/js/178.392431b3.js"><link rel="prefetch" href="/assets/js/179.31375329.js"><link rel="prefetch" href="/assets/js/18.e1886c50.js"><link rel="prefetch" href="/assets/js/180.eab449ae.js"><link rel="prefetch" href="/assets/js/181.de02433a.js"><link rel="prefetch" href="/assets/js/182.69f2a2ac.js"><link rel="prefetch" href="/assets/js/183.a7c29149.js"><link rel="prefetch" href="/assets/js/184.d172410d.js"><link rel="prefetch" href="/assets/js/185.55f5cb3b.js"><link rel="prefetch" href="/assets/js/186.fe28ccd6.js"><link rel="prefetch" href="/assets/js/187.7078cb3a.js"><link rel="prefetch" href="/assets/js/188.33d16ad2.js"><link rel="prefetch" href="/assets/js/189.429e8408.js"><link rel="prefetch" href="/assets/js/19.0c100fc8.js"><link rel="prefetch" href="/assets/js/190.9fc52549.js"><link rel="prefetch" href="/assets/js/191.579656e4.js"><link rel="prefetch" href="/assets/js/192.1faeacc0.js"><link rel="prefetch" href="/assets/js/193.26b1f939.js"><link rel="prefetch" href="/assets/js/194.7b06da95.js"><link rel="prefetch" href="/assets/js/195.99cc7e20.js"><link rel="prefetch" href="/assets/js/196.1ea3220f.js"><link rel="prefetch" href="/assets/js/197.3738ecaa.js"><link rel="prefetch" href="/assets/js/198.b5f0dcb0.js"><link rel="prefetch" href="/assets/js/199.4c37b475.js"><link rel="prefetch" href="/assets/js/20.d3ec4ffd.js"><link rel="prefetch" href="/assets/js/200.d0f3b502.js"><link rel="prefetch" href="/assets/js/201.56c48b30.js"><link rel="prefetch" href="/assets/js/202.dc7fe2ea.js"><link rel="prefetch" href="/assets/js/203.2e7e7f17.js"><link rel="prefetch" href="/assets/js/204.71fe20bf.js"><link rel="prefetch" href="/assets/js/205.41b84b9d.js"><link rel="prefetch" href="/assets/js/206.cdc16e15.js"><link rel="prefetch" href="/assets/js/21.718b7fbf.js"><link rel="prefetch" href="/assets/js/22.3ae17164.js"><link rel="prefetch" href="/assets/js/23.ad0a6015.js"><link rel="prefetch" href="/assets/js/24.4487bebb.js"><link rel="prefetch" href="/assets/js/25.41a5d823.js"><link rel="prefetch" href="/assets/js/26.4e51890f.js"><link rel="prefetch" href="/assets/js/27.fd478fd6.js"><link rel="prefetch" href="/assets/js/28.211d4aa6.js"><link rel="prefetch" href="/assets/js/29.5b060608.js"><link rel="prefetch" href="/assets/js/3.75220a5c.js"><link rel="prefetch" href="/assets/js/30.e24a6501.js"><link rel="prefetch" href="/assets/js/31.79b31301.js"><link rel="prefetch" href="/assets/js/32.743ba4bf.js"><link rel="prefetch" href="/assets/js/33.4fff5d91.js"><link rel="prefetch" href="/assets/js/34.ac036805.js"><link rel="prefetch" href="/assets/js/35.8021acce.js"><link rel="prefetch" href="/assets/js/36.469285e3.js"><link rel="prefetch" href="/assets/js/37.ea7464b1.js"><link rel="prefetch" href="/assets/js/38.be0f1986.js"><link rel="prefetch" href="/assets/js/39.42be918e.js"><link rel="prefetch" href="/assets/js/4.f5011ba2.js"><link rel="prefetch" href="/assets/js/40.ac571077.js"><link rel="prefetch" href="/assets/js/41.f593e342.js"><link rel="prefetch" href="/assets/js/42.bca3b705.js"><link rel="prefetch" href="/assets/js/43.f5070b9f.js"><link rel="prefetch" href="/assets/js/44.a3911525.js"><link rel="prefetch" href="/assets/js/45.b9fc883d.js"><link rel="prefetch" href="/assets/js/46.464c7e51.js"><link rel="prefetch" href="/assets/js/47.7807b8a1.js"><link rel="prefetch" href="/assets/js/48.d5ae16c3.js"><link rel="prefetch" href="/assets/js/49.6d24b07e.js"><link rel="prefetch" href="/assets/js/5.b582c580.js"><link rel="prefetch" href="/assets/js/50.4d1db70c.js"><link rel="prefetch" href="/assets/js/51.89065036.js"><link rel="prefetch" href="/assets/js/52.f308826a.js"><link rel="prefetch" href="/assets/js/53.f416c945.js"><link rel="prefetch" href="/assets/js/54.de986d56.js"><link rel="prefetch" href="/assets/js/55.a38a082e.js"><link rel="prefetch" href="/assets/js/56.1c42fbac.js"><link rel="prefetch" href="/assets/js/57.077a2b61.js"><link rel="prefetch" href="/assets/js/58.80d7cb5e.js"><link rel="prefetch" href="/assets/js/59.c2ee2cf4.js"><link rel="prefetch" href="/assets/js/6.52fc7c97.js"><link rel="prefetch" href="/assets/js/60.0bf8a1a7.js"><link rel="prefetch" href="/assets/js/61.1257ff7b.js"><link rel="prefetch" href="/assets/js/62.c41e8c91.js"><link rel="prefetch" href="/assets/js/63.39361554.js"><link rel="prefetch" href="/assets/js/64.597581ff.js"><link rel="prefetch" href="/assets/js/65.9e1a8d7c.js"><link rel="prefetch" href="/assets/js/66.334ddb5e.js"><link rel="prefetch" href="/assets/js/67.abb8980a.js"><link rel="prefetch" href="/assets/js/68.50d6ccbd.js"><link rel="prefetch" href="/assets/js/69.78adbe1e.js"><link rel="prefetch" href="/assets/js/7.069c08dd.js"><link rel="prefetch" href="/assets/js/70.b31703cd.js"><link rel="prefetch" href="/assets/js/71.ccd34bfc.js"><link rel="prefetch" href="/assets/js/72.a2477889.js"><link rel="prefetch" href="/assets/js/73.3c0a5fa3.js"><link rel="prefetch" href="/assets/js/74.e27d5da5.js"><link rel="prefetch" href="/assets/js/75.581bc1ad.js"><link rel="prefetch" href="/assets/js/76.1eb8f4e3.js"><link rel="prefetch" href="/assets/js/77.fdab5a7d.js"><link rel="prefetch" href="/assets/js/78.176ad3f1.js"><link rel="prefetch" href="/assets/js/79.a4feb4cd.js"><link rel="prefetch" href="/assets/js/8.5f8a95e5.js"><link rel="prefetch" href="/assets/js/80.74a9ea67.js"><link rel="prefetch" href="/assets/js/81.b475a94b.js"><link rel="prefetch" href="/assets/js/82.aa6cf5bc.js"><link rel="prefetch" href="/assets/js/83.3ef99db9.js"><link rel="prefetch" href="/assets/js/84.d93f9635.js"><link rel="prefetch" href="/assets/js/85.0e647886.js"><link rel="prefetch" href="/assets/js/86.dc078fed.js"><link rel="prefetch" href="/assets/js/87.975d4001.js"><link rel="prefetch" href="/assets/js/88.19f5ee6a.js"><link rel="prefetch" href="/assets/js/89.d6d84503.js"><link rel="prefetch" href="/assets/js/9.fc7a2252.js"><link rel="prefetch" href="/assets/js/90.6c1b5340.js"><link rel="prefetch" href="/assets/js/91.1f9f9747.js"><link rel="prefetch" href="/assets/js/92.d8d60a44.js"><link rel="prefetch" href="/assets/js/93.28110f57.js"><link rel="prefetch" href="/assets/js/94.f4c5cd90.js"><link rel="prefetch" href="/assets/js/95.afa19f49.js"><link rel="prefetch" href="/assets/js/96.abed696f.js"><link rel="prefetch" href="/assets/js/97.b4b650c9.js"><link rel="prefetch" href="/assets/js/98.7892a857.js"><link rel="prefetch" href="/assets/js/99.9c0aaa7c.js">
    <link rel="stylesheet" href="/assets/css/0.styles.e4d8b86b.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container no-sidebar"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <!----></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><!---->  <!----> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="exception-handling"><a href="#exception-handling" class="header-anchor">#</a> Exception Handling</h1> <blockquote><p>Failure is always a possibility.</p></blockquote> <p>Kotlin finds basic errors when it analyzes your program. Errors that cannot be detected at compile time must be dealt with at runtime. In [Exceptions](javascript:void(0)), you learned to throw exceptions. In this atom, we <em>catch</em> exceptions.</p> <p>Historically, failures were often disastrous. For example, programs written in the C language would simply stop working, lose their data, and potentially crash the operating system.</p> <p>Improved error handling is a powerful way to increase code reliability. Error handling is especially important when creating reusable program components. To create a robust system, each component must be robust. With consistent error handling, components can reliably communicate problems to client code.</p> <p>Modern applications often use concurrency, and a concurrent program must survive non-critical exceptions. A server, for example, should recover when an open session is terminated via an exception.</p> <p>Exceptions conflate three activities:</p> <ol><li>Error reporting</li> <li>Recovery</li> <li>Resource cleanup</li></ol> <p>Let’s consider each one.</p> <h3 id="reporting"><a href="#reporting" class="header-anchor">#</a> Reporting</h3> <p>Standard library exceptions are often adequate. For more specific exception handling, you can inherit new exception types from <code>Exception</code> or a subtype:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/DefiningExceptions.kt
package exceptionhandling
import atomictest.*

class Exception1(
  val value: Int
): Exception(&quot;wrong value: $value&quot;)

open class Exception2(
  description: String
): Exception(description)

class Exception3(
  description: String
): Exception2(description)

fun main() {
  capture {
    throw Exception1(13)
  } eq &quot;Exception1: wrong value: 13&quot;
  capture {
    throw Exception3(&quot;error&quot;)
  } eq &quot;Exception3: error&quot;
}
</code></pre></div><p>A <code>throw</code> expression, as in <code>main()</code>, requires an instance of a <code>Throwable</code> subtype. To define new exception types, inherit <code>Exception</code> (which extends <code>Throwable</code>). Both <code>Exception1</code> and <code>Exception2</code> inherit <code>Exception</code>, while <code>Exception3</code> inherits <code>Exception2</code>.</p> <h3 id="recovery"><a href="#recovery" class="header-anchor">#</a> Recovery</h3> <p>The ambition of exception handling is recovery. This means that you fix the problem, return the program to a stable state, and resume execution. Recovery often includes [logging](javascript:void(0)) information about the error.</p> <p>Quite often, recovery isn’t possible. An exception might represent an unrecoverable program failure, either a coding error or something uncontrollable in the environment.</p> <p>When an exception is thrown, the exception-handling mechanism looks for an appropriate place to continue execution. An exception keeps moving out to higher levels, from <code>function1()</code> that threw the exception, to <code>function2()</code> that calls <code>function1()</code>, to <code>function3()</code> that calls <code>function2()</code>, and so on until reaching <code>main()</code>. A matching handler <em>catches</em> the exception. This stops the search and runs that handler. If the program never finds a matching handler, it terminates with a console stack trace.</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/Stacktrace.kt
package stacktrace
import exceptionhandling.Exception1

fun function1(): Int =
  throw Exception1(-52)

fun function2() = function1()

fun function3() = function2()

fun main() {
//  function3()
}
</code></pre></div><p>Uncommenting the call to <code>function3()</code> produces the following stack trace:</p> <div class="language- extra-class"><pre class="language-text"><code>Exception in thread &quot;main&quot; exceptionhandling.Exception1: wrong value: -\
52
  at stacktrace.StacktraceKt.function1(Stacktrace.kt:6)
  at stacktrace.StacktraceKt.function2(Stacktrace.kt:8)
  at stacktrace.StacktraceKt.function3(Stacktrace.kt:10)
  at stacktrace.StacktraceKt.main(Stacktrace.kt:13)
  at stacktrace.StacktraceKt.main(Stacktrace.kt)
</code></pre></div><p>Any of <code>function1()</code>, <code>function2()</code> or <code>function3()</code> can <code>catch</code> the exception and handle it, preventing the exception from terminating the program.</p> <p>An <em>exception handler</em> is the <code>catch</code> keyword followed by a parameter list containing the exception you’re handling. This is followed by a block of code implementing the recovery.</p> <p>In the following example, the function <code>toss()</code> produces different exceptions for arguments 1-3, otherwise it returns “OK”. <code>test()</code> contains a complete set of handlers for the <code>toss()</code> function:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/Handlers.kt
package exceptionhandling
import atomictest.eq

fun toss(which: Int) = when (which) {
  1 -&gt; throw Exception1(1)
  2 -&gt; throw Exception2(&quot;Exception 2&quot;)
  3 -&gt; throw Exception3(&quot;Exception 3&quot;)
  else -&gt; &quot;OK&quot;
}

fun test(which: Int): Any? =
  try {
    toss(which)
  } catch (e: Exception1) {
    e.value
  } catch (e: Exception3) {
    e.message
  } catch (e: Exception2) {
    e.message
  }

fun main() {
  test(0) eq &quot;OK&quot;
  test(1) eq 1
  test(2) eq &quot;Exception 2&quot;
  test(3) eq &quot;Exception 3&quot;
}
</code></pre></div><p>When you call <code>toss()</code> you must <code>catch</code> all relevant <code>toss()</code> exceptions, allowing non-relevant exceptions to “bubble up” and be caught elsewhere.</p> <p>The entire <code>try</code>-<code>catch</code> in <code>test()</code> is a single expression: it returns either the last expression of the <code>try</code> body or the last expression of the <code>catch</code> clause matching an exception. If no <code>catch</code> handles the exception, that exception is thrown further up the stack. If uncaught, it generates a stack trace.</p> <p>Because <code>Exception3</code> extends <code>Exception2</code>, an <code>Exception3</code> is handled as an <code>Exception2</code> if <code>Exception2</code>’s <code>catch</code> appears in the sequence of handlers before <code>Exception3</code>’s <code>catch</code>:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/Hierarchy.kt
package exceptionhandling
import atomictest.eq

fun testCatchOrder(which: Int) =
  try {
    toss(which)
  } catch (e: Exception2) {    // [1]
    &quot;Handler for Exception2 got ${e.message}&quot;
  } catch (e: Exception3) {    // [2]
    &quot;Handler for Exception3 got ${e.message}&quot;
  }

fun main() {
  testCatchOrder(2) eq
    &quot;Handler for Exception2 got Exception 2&quot;
  testCatchOrder(3) eq
    &quot;Handler for Exception2 got Exception 3&quot;
}
</code></pre></div><p>The <code>catch</code>-clause order means an <code>Exception3</code> is caught by line <strong>[1]</strong>, despite the more specific type of exception handler in line <strong>[2]</strong>.</p> <h4 id="exception-subtypes"><a href="#exception-subtypes" class="header-anchor">#</a> Exception Subtypes</h4> <p>In <code>testCode()</code>, an incorrect <code>code</code> argument throws an <code>IllegalArgumentException</code>:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/LibraryException.kt
package exceptionhandling
import atomictest.*

fun testCode(code: Int) {
  if (code &lt;= 1000) {
    throw IllegalArgumentException(
      &quot;'code' must be &gt; 1000: $code&quot;)
  }
}

fun main() {
  try {
    // A1 is 161 in base-16 (hex) notation:
    testCode(&quot;A1&quot;.toInt(16))
  } catch (e: IllegalArgumentException) {
    e.message eq
      &quot;'code' must be &gt; 1000: 161&quot;
  }
  try {
    testCode(&quot;0&quot;.toInt(1))
  } catch (e: IllegalArgumentException) {
    e.message eq
      &quot;radix 1 was not in valid range 2..36&quot;
  }
}
</code></pre></div><p>An <code>IllegalArgumentException</code> is thrown by both <code>testCode()</code> and the library function <code>toInt(radix)</code>. This results in the somewhat confusing error messages in <code>main()</code>. The problem is that we are using the same exception to represent two different issues. We solve it by throwing a new exception type called <code>IncorrectInputException</code> for our error:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/NewException.kt
package exceptionhandling
import atomictest.eq

class IncorrectInputException(
  message: String
): Exception(message)

fun checkCode(code: Int) {
  if (code &lt;= 1000) {
    throw IncorrectInputException(
      &quot;Code must be &gt; 1000: $code&quot;)
  }
}

fun main() {
  try {
    checkCode(&quot;A1&quot;.toInt(16))
  } catch (e: IncorrectInputException) {
    e.message eq &quot;Code must be &gt; 1000: 161&quot;
  } catch (e: IllegalArgumentException) {
    &quot;Produces error&quot; eq &quot;if it gets here&quot;
  }
  try {
    checkCode(&quot;1&quot;.toInt(1))
  } catch (e: IncorrectInputException) {
    &quot;Produces error&quot; eq &quot;if it gets here&quot;
  } catch (e: IllegalArgumentException) {
    e.message eq
      &quot;radix 1 was not in valid range 2..36&quot;
  }
}
</code></pre></div><p>Now each issue has its own handler.</p> <p>Resist creating too many exception types. As a rule of thumb, use different exception types to distinguish different handling schemes, and use different constructor parameters to provide details for a particular handling scheme.</p> <h3 id="resource-cleanup"><a href="#resource-cleanup" class="header-anchor">#</a> Resource Cleanup</h3> <p>When failure is inevitable, automatic resource cleanup helps other parts of the program to continue running safely.</p> <p><code>finally</code> ensures resource cleanup during exception handling. A <code>finally</code> clause always runs, regardless of whether you leave a <code>try</code> block normally or exceptionally:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/TryFinally.kt
package exceptionhandling
import atomictest.*

fun checkValue(value: Int) {
  try {
    trace(value)
    if (value &lt;= 0)
      throw IllegalArgumentException(
        &quot;value must be positive: $value&quot;)
  } finally {
    trace(&quot;In finally clause for $value&quot;)
  }
}

fun main() {
  listOf(10, -10).forEach {
    try {
      checkValue(it)
    } catch (e: IllegalArgumentException) {
      trace(&quot;In catch clause for main()&quot;)
      trace(e.message)
    }
  }
  trace eq &quot;&quot;&quot;
    10
    In finally clause for 10
    -10
    In finally clause for -10
    In catch clause for main()
    value must be positive: -10
  &quot;&quot;&quot;
}
</code></pre></div><p><code>finally</code> works even with intermediate <code>catch</code> clauses. For example, suppose a switch must be turned off when you’re done with it:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/GuaranteedCleanup.kt
package exceptionhandling
import atomictest.eq

data class Switch(
  var on: Boolean = false,
  var result: String = &quot;OK&quot;
)

fun testFinally(i: Int): Switch {
  val sw = Switch()
  try {
    sw.on = true
    when (i) {
      0 -&gt; throw IllegalStateException()
      1 -&gt; return sw                 // [1]
    }
  } catch (e: IllegalStateException) {
    sw.result = &quot;exception&quot;
  } finally {
    sw.on = false
  }
  return sw
}

fun main() {
  testFinally(0) eq
    &quot;Switch(on=false, result=exception)&quot;
  testFinally(1) eq
    &quot;Switch(on=false, result=OK)&quot;    // [2]
  testFinally(2) eq
    &quot;Switch(on=false, result=OK)&quot;
}
</code></pre></div><p>Even if we <code>return</code> inside a <code>try</code> (<strong>[1]</strong>), the <code>finally</code> clause still runs (<strong>[2]</strong>). Whether <code>testFinally()</code> completes normally or with an exception, the <code>finally</code> clause always executes.</p> <h3 id="exception-handling-in-atomictest"><a href="#exception-handling-in-atomictest" class="header-anchor">#</a> Exception Handling in AtomicTest</h3> <p>This book uses AtomicTest’s <code>capture()</code> to ensure that expected exceptions are thrown. <code>capture()</code> takes a function argument and returns a <code>CapturedException</code> object containing the exception class and error message:</p> <div class="language- extra-class"><pre class="language-text"><code>// ExceptionHandling/CaptureImplementation.kt
package exceptionhandling
import atomictest.CapturedException

fun capture(f:() -&gt; Unit): CapturedException =
  try {                                 // [1]
    f()
    CapturedException(null,
      &quot;&lt;Error&gt;: Expected an exception&quot;) // [2]
  } catch (e: Throwable) {              // [3]
    CapturedException(e::class,         // [4]
      if (e.message != null) &quot;: ${e.message}&quot;
      else &quot;&quot;)
  }

fun main() {
  capture {
    throw Exception(&quot;!!!&quot;)
  } eq &quot;Exception: !!!&quot;                 // [5]
  capture {
    1
  } eq &quot;&lt;Error&gt;: Expected an exception&quot;
}
</code></pre></div><p><code>capture()</code> calls its function argument <code>f</code> within a <code>try</code> block (<strong>[1]</strong>), handling all possible exceptions by catching <code>Throwable</code> (<strong>[3]</strong>). If no exception is thrown, the <code>CapturedException</code> message indicates that an exception was expected (<strong>[2]</strong>). If an exception is caught, the returned <code>CapturedException</code> contains the exception class and a message (<strong>[4]</strong>). A <code>CapturedException</code> can be compared to a <code>String</code> using <code>eq</code> (<strong>[5]</strong>).</p> <p>Ordinarily you won’t catch <code>Throwable</code>, but will process each specific exception type.</p> <h3 id="guidelines"><a href="#guidelines" class="header-anchor">#</a> Guidelines</h3> <p>Recovering from exceptions turns out to be remarkably rare, considering that recovery was the original intent. The primary purpose of exceptions in Kotlin is to discover program bugs, not recovery. Catching exceptions in ordinary Kotlin code is thus a “code smell.”</p> <p>Here are guidelines for programming with exceptions in Kotlin:</p> <ol><li><p><strong>Logic Errors</strong>: These are bugs in your code. Either don’t catch them at all (and produce a stack trace), or catch them at the top level of your application and report the bugs, possibly restarting the affected operation.</p></li> <li><p>Data Errors</p> <p>: These are errors from bad data that the programmer cannot control. The application must somehow deal with the problem without blaming it on program logic. For example, we’ve used</p> <div class="language- extra-class"><pre class="language-text"><code>String.toInt()
</code></pre></div><p>this atom, which throws an exception for an inappropriate</p> <div class="language- extra-class"><pre class="language-text"><code>String
</code></pre></div><p>. It also has a companion</p> <div class="language- extra-class"><pre class="language-text"><code>String.toIntOrNull()
</code></pre></div><p>that produces a</p> <div class="language- extra-class"><pre class="language-text"><code>null
</code></pre></div><p>upon failure so you can use it in an expression such as</p> <div class="language- extra-class"><pre class="language-text"><code>val n = string.toIntOrNull() ?: default
</code></pre></div><p>.</p> <p>The Kotlin library is designed around dealing with a bad result by returning a <code>null</code> instead of throwing an exception. Operations that are expected to occasionally fail will usually have an “OrNull” version that you can use instead of the exception version.</p></li> <li><p>[<strong>Check instructions</strong>](javascript:void(0)) test for logic errors. These produce exceptions when they find a bug, but they look like function calls so you don’t explicitly throw exceptions in your code.</p></li> <li><p><strong>Input/Output Errors</strong>: These are external conditions that you can’t control and you can’t ignore. However, using the “OrNull” approach rapidly obscures the understandability of the code. More importantly, you often <em>can</em> recover from I/O errors, typically by retrying the operation. Thus, I/O operations in Kotlin throw exceptions, so you’ll have code in your applications that handle those and attempt to recover from them.</p></li></ol> <p><em><strong>Exercises and solutions can be found at www.AtomicKotlin.com.</strong></em></p></div> <footer class="page-edit"><!----> <!----></footer> <!----> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.885421c5.js" defer></script><script src="/assets/js/2.63033d72.js" defer></script><script src="/assets/js/102.ad03aa52.js" defer></script>
  </body>
</html>
